File: 0_0100/0003_longuest_substring_without_repeating_characters.rs
use std::cmp::max;
use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut occurs = HashMap::new();
        let (mut res, mut before, mut cur) = (0, -1, 0);
        
        s.chars()
            .for_each(|c| {     
                match occurs.insert(c, cur) {
                    Some(x) => before = max(before, x),
                    None => {}
                }
                res = max(res, cur - before);
                cur += 1;
            }); 
        
        res
    }
}

File: 0001_0100/0003_longuest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> occurs;
        int res = 0, before = -1, cur = 0;

        for (char c : s) {
            if (occurs.count(c) == 1) {
                before = max(before, occurs[c]);
            }
            occurs[c] = cur;
            res = max(res, cur - before);
            cur++;
        }
        return res;
    }
};